File: backend/src/test/java/com/back/ourlog/domain/statistics/contoroller/StatisticsControllerTest.java
package com.back.ourlog.domain.statistics.contoroller;

import com.back.ourlog.domain.statistics.controller.StatisticsController;
import com.back.ourlog.domain.statistics.dto.*;
import com.back.ourlog.domain.statistics.enums.PeriodOption;
import com.back.ourlog.domain.statistics.service.StatisticsService;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.MediaType;
import org.springframework.security.test.context.support.WithUserDetails;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.ResultActions;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.handler;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@ActiveProfiles("test")
@SpringBootTest
@Transactional
@AutoConfigureMockMvc
public class StatisticsControllerTest {

    @Autowired
    private MockMvc mvc;

    @Autowired
    private StatisticsService statisticsService;

    @Test
    @DisplayName("통계 카드 조회")
    @WithUserDetails("user1@test.com")
    void 통계_카드_조회() throws Exception {

        ResultActions resultActions = mvc.perform(
                get("/api/v1/statistics/card")
        ).andDo(print());

        resultActions
                .andExpect(handler().handlerType(StatisticsController.class))
                .andExpect(handler().methodName("getStatisticsCard"))
                .andExpect(status().isOk());

        StatisticsCardDto statisticsCardDto = statisticsService.getStatisticsCardByUserId(1);

        resultActions
                .andExpect(jsonPath("$.totalDiaryCount").value(statisticsCardDto.getTotalDiaryCount()))
                .andExpect(jsonPath("$.averageRating").value(statisticsCardDto.getAverageRating()))
                .andExpect(jsonPath("$.favoriteType").value(statisticsCardDto.getFavoriteType()))
                .andExpect(jsonPath("$.favoriteTypeCount").value(statisticsCardDto.getFavoriteTypeCount()))
                .andExpect(jsonPath("$.favoriteEmotion").value(statisticsCardDto.getFavoriteEmotion()))
                .andExpect(jsonPath("$.favoriteEmotionCount").value(statisticsCardDto.getFavoriteEmotionCount()));

        System.out.println("Total Diary Count: " + statisticsCardDto.getTotalDiaryCount());
        System.out.println("Average Rating: " + statisticsCardDto.getAverageRating());
        System.out.println("Favorite Type: " + statisticsCardDto.getFavoriteType());
        System.out.println("Favorite Type Count: " + statisticsCardDto.getFavoriteTypeCount());
        System.out.println("Favorite Emotion: " + statisticsCardDto.getFavoriteEmotion());
        System.out.println("Favorite Emotion Count: " + statisticsCardDto.getFavoriteEmotionCount());

    }

    @Test
    @DisplayName("최근 6개월 월 별 감상 수 조회")
    @WithUserDetails("user1@test.com")
    void 최근_6개월_월_별_감상_수_조회() throws Exception {

        ResultActions resultActions = mvc.perform(
                get("/api/v1/statistics/monthly-diary-graph")
        ).andDo(print());

        resultActions
                .andExpect(handler().handlerType(StatisticsController.class))
                .andExpect(handler().methodName("getLast6MonthsDiaryCounts"))
                .andExpect(status().isOk());

        List<MonthlyDiaryCount> monthlyDiaryCounts = statisticsService.getLast6MonthsDiaryCountsByUser(1);

        for(int i = 0; i < monthlyDiaryCounts.size(); i++) {
            MonthlyDiaryCount monthlyDiaryCount = monthlyDiaryCounts.get(i);
            System.out.println("Period: " + monthlyDiaryCount.getPeriod() + ", Views: " + monthlyDiaryCount.getViews());
            resultActions
                    .andExpect(jsonPath("$.[%d].period".formatted(i)).value(monthlyDiaryCount.getPeriod()))
                    .andExpect(jsonPath("$.[%d].views".formatted(i)).value(monthlyDiaryCount.getViews()));
        }
    }

    @Test
    @DisplayName("콘텐츠 타입 분포 조회")
    @WithUserDetails("user1@test.com")
    void 콘텐츠_타입_분포_조회() throws Exception {

        ResultActions resultActions = mvc.perform(
                get("/api/v1/statistics/type-distribution")
        ).andDo(print());

        resultActions
                .andExpect(handler().handlerType(StatisticsController.class))
                .andExpect(handler().methodName("getTypeDistribution"))
                .andExpect(status().isOk());

        List<TypeCountDto> typeCountDtos = statisticsService.getTypeDistributionByUser(1);

        for(int i = 0; i < typeCountDtos.size(); i++) {
            TypeCountDto typeCountDto = typeCountDtos.get(i);
            System.out.println("Type: " + typeCountDto.getType() + ", Count: " + typeCountDto.getCount());
            resultActions
                    .andExpect(jsonPath("$.[%d].type".formatted(i)).value(typeCountDto.getType()))
                    .andExpect(jsonPath("$.[%d].count".formatted(i)).value(typeCountDto.getCount()));
        }
    }

    @Test
    @DisplayName("타입 그래프 조회")
    @WithUserDetails("user1@test.com")
    void 타입_그래프() throws Exception {

        ResultActions resultActions = mvc.perform(
                get("/api/v1/statistics/type-graph")
                .param("period", "ALL")  // 예시로 MONTH 기간을 사용
        ).andDo(print());

        resultActions
                .andExpect(handler().handlerType(StatisticsController.class))
                .andExpect(handler().methodName("getTypeGraph"))
                .andExpect(status().isOk());

        TypeGraphResponse typeGraphResponse = statisticsService.getTypeGraph(new TypeGraphRequest(1, PeriodOption.ALL));

        List<TypeLineGraphDto> trend = typeGraphResponse.getTypeLineGraph();
        List<TypeRankDto> ranking = typeGraphResponse.getTypeRanking();

        for(int i = 0; i < trend.size(); i++) {
            TypeLineGraphDto typeLineGraphDto = trend.get(i);
            System.out.println("Axis Label: " + typeLineGraphDto.getAxisLabel() + ", Type: " + typeLineGraphDto.getType() + ", Count: " + typeLineGraphDto.getCount());
            resultActions
                    .andExpect(jsonPath("$.typeLineGraph.[%d].axisLabel".formatted(i)).value(typeLineGraphDto.getAxisLabel()))
                    .andExpect(jsonPath("$.typeLineGraph.[%d].type".formatted(i)).value(typeLineGraphDto.getType().name()))
                    .andExpect(jsonPath("$.typeLineGraph.[%d].count".formatted(i)).value(typeLineGraphDto.getCount()));
        }
        for(int i = 0; i < ranking.size(); i++) {
            TypeRankDto typeRankDto = ranking.get(i);
            System.out.println("Type: " + typeRankDto.getType() + ", Count: " + typeRankDto.getTotalCount());
            resultActions
                    .andExpect(jsonPath("$.typeRanking.[%d].type".formatted(i)).value(typeRankDto.getType().name()))
                    .andExpect(jsonPath("$.typeRanking.[%d].totalCount".formatted(i)).value(typeRankDto.getTotalCount()));
        }

    }

    @Test
    @DisplayName("장르 그래프 조회")
    @WithUserDetails("user1@test.com")
    void 장르_그래프() throws Exception {

        ResultActions resultActions = mvc.perform(
                get("/api/v1/statistics/genre-graph")
                        .param("period", "ALL")  // 예시로 MONTH 기간을 사용
        ).andDo(print());

        resultActions
                .andExpect(handler().handlerType(StatisticsController.class))
                .andExpect(handler().methodName("getGenreGraph"))
                .andExpect(status().isOk());

        GenreGraphResponse genreGraphResponse = statisticsService.getGenreGraph(1, PeriodOption.ALL);

        List<GenreLineGraphDto> graph = genreGraphResponse.getGenreLineGraph();
        List<GenreRankDto> ranking = genreGraphResponse.getGenreRanking();

        for(int i = 0; i < graph.size(); i++) {
            GenreLineGraphDto genreLineGraphDto = graph.get(i);
            System.out.println("Axis Label: " + genreLineGraphDto.getAxisLabel() + ", Genre: " + genreLineGraphDto.getGenre() + ", Count: " + genreLineGraphDto.getCount());
            resultActions
                    .andExpect(jsonPath("$.genreLineGraph.[%d].axisLabel".formatted(i)).value(genreLineGraphDto.getAxisLabel()))
                    .andExpect(jsonPath("$.genreLineGraph.[%d].genre".formatted(i)).value(genreLineGraphDto.getGenre()))
                    .andExpect(jsonPath("$.genreLineGraph.[%d].count".formatted(i)).value(genreLineGraphDto.getCount()));
        }

        for(int i = 0; i < ranking.size(); i++) {
            GenreRankDto genreRankDto = ranking.get(i);
            System.out.println("Genre: " + genreRankDto.getGenre() + ", Count: " + genreRankDto.getTotalCount());
            resultActions
                    .andExpect(jsonPath("$.genreRanking.[%d].genre".formatted(i)).value(genreRankDto.getGenre()))
                    .andExpect(jsonPath("$.genreRanking.[%d].totalCount".formatted(i)).value(genreRankDto.getTotalCount()));
        }
    }

    @Test
    @DisplayName("감정 그래프 조회")
    @WithUserDetails("user1@test.com")
    void 감정_그래프() throws Exception {

        ResultActions resultActions = mvc.perform(
                get("/api/v1/statistics/emotion-graph")
                        .param("period", "ALL")  // 예시로 MONTH 기간을 사용
        ).andDo(print());

        resultActions
                .andExpect(handler().handlerType(StatisticsController.class))
                .andExpect(handler().methodName("getEmotionGraph"))
                .andExpect(status().isOk());

        EmotionGraphResponse emotionGraphResponse = statisticsService.getEmotionGraph(1, PeriodOption.ALL);

        List<EmotionLineGraphDto> line = emotionGraphResponse.getEmotionLineGraph();
        List<EmotionRankDto> ranking = emotionGraphResponse.getEmotionRanking();

        for(int i = 0; i < line.size(); i++) {
            EmotionLineGraphDto emotionLineGraphDto = line.get(i);
            System.out.println("Axis Label: " + emotionLineGraphDto.getAxisLabel() + ", Emotion: " + emotionLineGraphDto.getEmotion() + ", Count: " + emotionLineGraphDto.getCount());
            resultActions
                    .andExpect(jsonPath("$.emotionLineGraph.[%d].axisLabel".formatted(i)).value(emotionLineGraphDto.getAxisLabel()))
                    .andExpect(jsonPath("$.emotionLineGraph.[%d].emotion".formatted(i)).value(emotionLineGraphDto.getEmotion()))
                    .andExpect(jsonPath("$.emotionLineGraph.[%d].count".formatted(i)).value(emotionLineGraphDto.getCount()));
        }
        for(int i = 0; i < ranking.size(); i++) {
            EmotionRankDto emotionRankDto = ranking.get(i);
            System.out.println("Emotion: " + emotionRankDto.getEmotion() + ", Count: " + emotionRankDto.getTotalCount());
            resultActions
                    .andExpect(jsonPath("$.emotionRanking.[%d].emotion".formatted(i)).value(emotionRankDto.getEmotion()))
                    .andExpect(jsonPath("$.emotionRanking.[%d].totalCount".formatted(i)).value(emotionRankDto.getTotalCount()));
        }
    }

    @Test
    @DisplayName("ott 그래프 조회")
    @WithUserDetails("user1@test.com")
    void ott_그래프() throws Exception {

        ResultActions resultActions = mvc.perform(
                get("/api/v1/statistics/ott-graph")
                        .param("period", "ALL")  // 예시로 MONTH 기간을 사용
        ).andDo(print());

        resultActions
                .andExpect(handler().handlerType(StatisticsController.class))
                .andExpect(handler().methodName("getOttGraph"))
                .andExpect(status().isOk());

        OttGraphResponse ottGraphResponse = statisticsService.getOttGraph(1, PeriodOption.ALL);

        List<OttLineGraphDto> line = ottGraphResponse.getOttLineGraph();
        List<OttRankDto> ranking = ottGraphResponse.getOttRanking();

        for(int i = 0; i < line.size(); i++) {
            OttLineGraphDto ottLineGraphDto = line.get(i);
            System.out.println("Axis Label: " + ottLineGraphDto.getAxisLabel() + ", OTT: " + ottLineGraphDto.getOttName() + ", Count: " + ottLineGraphDto.getCount());
            resultActions
                    .andExpect(jsonPath("$.ottLineGraph.[%d].axisLabel".formatted(i)).value(ottLineGraphDto.getAxisLabel()))
                    .andExpect(jsonPath("$.ottLineGraph.[%d].ottName".formatted(i)).value(ottLineGraphDto.getOttName()))
                    .andExpect(jsonPath("$.ottLineGraph.[%d].count".formatted(i)).value(ottLineGraphDto.getCount()));
        }

        for(int i = 0; i < ranking.size(); i++) {
            OttRankDto ottRankDto = ranking.get(i);
            System.out.println("OTT: " + ottRankDto.getOttName() + ", Count: " + ottRankDto.getTotalCount());
            resultActions
                    .andExpect(jsonPath("$.ottRanking.[%d].ottName".formatted(i)).value(ottRankDto.getOttName()))
                    .andExpect(jsonPath("$.ottRanking.[%d].totalCount".formatted(i)).value(ottRankDto.getTotalCount()));
        }
    }
}

File: backend/src/test/java/com/back/ourlog/domain/statistics/controller/StatisticsControllerTest.kt
package com.back.ourlog.domain.statistics.controller

import com.back.ourlog.domain.statistics.enums.PeriodOption
import com.back.ourlog.domain.statistics.service.StatisticsService
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.security.test.context.support.WithUserDetails
import org.springframework.test.context.ActiveProfiles
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.get
import org.springframework.test.web.servlet.result.MockMvcResultMatchers
import org.springframework.transaction.annotation.Transactional

@ActiveProfiles("test")
@SpringBootTest
@Transactional
@AutoConfigureMockMvc
class StatisticsControllerTest @Autowired constructor(
    private val mvc: MockMvc,
    private val statisticsService: StatisticsService
) {

    @Test
    @DisplayName("통계 카드 조회")
    @WithUserDetails("user1@test.com")
    fun `통계 카드 조회`() {
        val resultActions = mvc.get("/api/v1/statistics/card").andDo { print() }

        resultActions
            .andExpect { MockMvcResultMatchers.handler().handlerType(StatisticsController::class.java) }
            .andExpect { MockMvcResultMatchers.handler().methodName("getStatisticsCard") }
            .andExpect { MockMvcResultMatchers.status().isOk }

        val statisticsCardDto = statisticsService.getStatisticsCardByUserId(1)

        resultActions
            .andExpect { MockMvcResultMatchers.jsonPath("$.data.totalDiaryCount").value(statisticsCardDto.totalDiaryCount) }
            .andExpect { MockMvcResultMatchers.jsonPath("$.data.averageRating").value(statisticsCardDto.averageRating) }
            .andExpect { MockMvcResultMatchers.jsonPath("$.data.favoriteTypeAndCountDto.favoriteType").value(statisticsCardDto.favoriteTypeAndCountDto.favoriteType) }
            .andExpect { MockMvcResultMatchers.jsonPath("$.data.favoriteTypeAndCountDto.favoriteTypeCount").value(statisticsCardDto.favoriteTypeAndCountDto.favoriteTypeCount) }
            .andExpect { MockMvcResultMatchers.jsonPath("$.data.favoriteEmotionAndCountDto.favoriteEmotion").value(statisticsCardDto.favoriteEmotionAndCountDto.favoriteEmotion)}
            .andExpect { MockMvcResultMatchers.jsonPath("$.data.favoriteEmotionAndCountDto.favoriteEmotionCount").value(statisticsCardDto.favoriteEmotionAndCountDto.favoriteEmotionCount) }

        println("Total Diary Count: ${statisticsCardDto.totalDiaryCount}")
        println("Average Rating: ${statisticsCardDto.averageRating}")
        println("Favorite Type: ${statisticsCardDto.favoriteTypeAndCountDto.favoriteType}")
        println("Favorite Type Count: ${statisticsCardDto.favoriteTypeAndCountDto.favoriteTypeCount}")
        println("Favorite Emotion: ${statisticsCardDto.favoriteEmotionAndCountDto.favoriteEmotion}")
        println("Favorite Emotion Count: ${statisticsCardDto.favoriteEmotionAndCountDto.favoriteEmotionCount}")
    }

    @Test
    @DisplayName("최근 6개월 월 별 감상 수 조회")
    @WithUserDetails("user1@test.com")
    fun `최근 6개월 월 별 감상 수 조회`() {
        val resultActions = mvc.get("/api/v1/statistics/monthly-diary-graph").andDo { print() }

        resultActions
            .andExpect { MockMvcResultMatchers.handler().handlerType(StatisticsController::class.java) }
            .andExpect { MockMvcResultMatchers.handler().methodName("getLast6MonthsDiaryCounts") }
            .andExpect { MockMvcResultMatchers.status().isOk }

        val monthlyDiaryCounts = statisticsService.getLast6MonthsDiaryCountsByUser(1)

        monthlyDiaryCounts.forEachIndexed { i, monthlyDiaryCount ->
            println("Period: ${monthlyDiaryCount.period}, Views: ${monthlyDiaryCount.views}")
            resultActions
                .andExpect { MockMvcResultMatchers.jsonPath("$.data.[$i].period").value(monthlyDiaryCount.period) }
                .andExpect { MockMvcResultMatchers.jsonPath("$.data.[$i].views").value(monthlyDiaryCount.views) }
        }
    }

    @Test
    @DisplayName("콘텐츠 타입 분포 조회")
    @WithUserDetails("user1@test.com")
    fun `콘텐츠 타입 분포 조회`() {
        val resultActions = mvc.get("/api/v1/statistics/type-distribution").andDo { print() }

        resultActions
            .andExpect { MockMvcResultMatchers.handler().handlerType(StatisticsController::class.java) }
            .andExpect { MockMvcResultMatchers.handler().methodName("getTypeDistribution") }
            .andExpect { MockMvcResultMatchers.status().isOk }

        val typeCountDtos = statisticsService.getTypeDistributionByUser(1)

        typeCountDtos.forEachIndexed { i, typeCountDto ->
            println("Type: ${typeCountDto.type}, Count: ${typeCountDto.count}")
            resultActions
                .andExpect { MockMvcResultMatchers.jsonPath("$.data.[$i].type").value(typeCountDto.type) }
                .andExpect { MockMvcResultMatchers.jsonPath("$.data.[$i].count").value(typeCountDto.count) }
        }
    }

    @Test
    @DisplayName("타입 그래프 조회")
    @WithUserDetails("user1@test.com")
    fun `타입 그래프`() {
        val resultActions = mvc.get("/api/v1/statistics/type-graph") {
            param("period", "ALL")
        }.andDo { print() }

        resultActions
            .andExpect { MockMvcResultMatchers.handler().handlerType(StatisticsController::class.java) }
            .andExpect { MockMvcResultMatchers.handler().methodName("getTypeGraph") }
            .andExpect { MockMvcResultMatchers.status().isOk }

        val typeGraphResponse = statisticsService.getTypeGraph(1, PeriodOption.ALL)

        val trend = typeGraphResponse.typeLineGraph
        val ranking = typeGraphResponse.typeRanking

        trend.forEachIndexed { i, typeLineGraphDto ->
            println("Axis Label: ${typeLineGraphDto.axisLabel}, Type: ${typeLineGraphDto.type}, Count: ${typeLineGraphDto.count}")
            resultActions
                .andExpect { MockMvcResultMatchers.jsonPath("$.data.typeLineGraph.[$i].axisLabel").value(typeLineGraphDto.axisLabel) }
                .andExpect { MockMvcResultMatchers.jsonPath("$.data.typeLineGraph.[$i].type").value(typeLineGraphDto.type.name) }
                .andExpect { MockMvcResultMatchers.jsonPath("$.data.typeLineGraph.[$i].count").value(typeLineGraphDto.count) }
        }
        ranking.forEachIndexed { i, typeRankDto ->
            println("Type: ${typeRankDto.type}, Count: ${typeRankDto.totalCount}")
            resultActions
                .andExpect { MockMvcResultMatchers.jsonPath("$.data.typeRanking.[$i].type").value(typeRankDto.type.name) }
                .andExpect { MockMvcResultMatchers.jsonPath("$.data.typeRanking.[$i].totalCount").value(typeRankDto.totalCount) }
        }
    }

    @Test
    @DisplayName("장르 그래프 조회")
    @WithUserDetails("user1@test.com")
    fun `장르 그래프`() {
        val resultActions = mvc.get("/api/v1/statistics/genre-graph") {
            param("period", "ALL")
        }.andDo { print() }

        resultActions
            .andExpect { MockMvcResultMatchers.handler().handlerType(StatisticsController::class.java) }
            .andExpect { MockMvcResultMatchers.handler().methodName("getGenreGraph") }
            .andExpect { MockMvcResultMatchers.status().isOk }

        val genreGraphResponse = statisticsService.getGenreGraph(1, PeriodOption.ALL)

        val graph = genreGraphResponse.genreLineGraph
        val ranking = genreGraphResponse.genreRanking

        graph.forEachIndexed { i, genreLineGraphDto ->
            println("Axis Label: ${genreLineGraphDto.axisLabel}, Genre: ${genreLineGraphDto.genre}, Count: ${genreLineGraphDto.count}")
            resultActions
                .andExpect { MockMvcResultMatchers.jsonPath("$.data.genreLineGraph.[$i].axisLabel").value(genreLineGraphDto.axisLabel) }
                .andExpect { MockMvcResultMatchers.jsonPath("$.data.genreLineGraph.[$i].genre").value(genreLineGraphDto.genre) }
                .andExpect { MockMvcResultMatchers.jsonPath("$.data.genreLineGraph.[$i].count").value(genreLineGraphDto.count) }
        }

        ranking.forEachIndexed { i, genreRankDto ->
            println("Genre: ${genreRankDto.genre}, Count: ${genreRankDto.totalCount}")
            resultActions
                .andExpect { MockMvcResultMatchers.jsonPath("$.data.genreRanking.[$i].genre").value(genreRankDto.genre) }
                .andExpect { MockMvcResultMatchers.jsonPath("$.data.genreRanking.[$i].totalCount").value(genreRankDto.totalCount) }
        }
    }

    @Test
    @DisplayName("감정 그래프 조회")
    @WithUserDetails("user1@test.com")
    fun `감정 그래프`() {
        val resultActions = mvc.get("/api/v1/statistics/emotion-graph") {
            param("period", "ALL")
        }.andDo { print() }

        resultActions
            .andExpect { MockMvcResultMatchers.handler().handlerType(StatisticsController::class.java) }
            .andExpect { MockMvcResultMatchers.handler().methodName("getEmotionGraph") }
            .andExpect { MockMvcResultMatchers.status().isOk }

        val emotionGraphResponse = statisticsService.getEmotionGraph(1, PeriodOption.ALL)

        val line = emotionGraphResponse.emotionLineGraph
        val ranking = emotionGraphResponse.emotionRanking

        line.forEachIndexed { i, emotionLineGraphDto ->
            println("Axis Label: ${emotionLineGraphDto.axisLabel}, Emotion: ${emotionLineGraphDto.emotion}, Count: ${emotionLineGraphDto.count}")
            resultActions
                .andExpect { MockMvcResultMatchers.jsonPath("$.data.emotionLineGraph.[$i].axisLabel").value(emotionLineGraphDto.axisLabel) }
                .andExpect { MockMvcResultMatchers.jsonPath("$.data.emotionLineGraph.[$i].emotion").value(emotionLineGraphDto.emotion) }
                .andExpect { MockMvcResultMatchers.jsonPath("$.data.emotionLineGraph.[$i].count").value(emotionLineGraphDto.count) }
        }
        ranking.forEachIndexed { i, emotionRankDto ->
            println("Emotion: ${emotionRankDto.emotion}, Count: ${emotionRankDto.totalCount}")
            resultActions
                .andExpect { MockMvcResultMatchers.jsonPath("$.data.emotionRanking.[$i].emotion").value(emotionRankDto.emotion) }
                .andExpect { MockMvcResultMatchers.jsonPath("$.data.emotionRanking.[$i].totalCount").value(emotionRankDto.totalCount) }
        }
    }

    @Test
    @DisplayName("ott 그래프 조회")
    @WithUserDetails("user1@test.com")
    fun `ott 그래프`() {
        val resultActions = mvc.get("/api/v1/statistics/ott-graph") {
            param("period", "ALL")
        }.andDo { print() }

        resultActions
            .andExpect { MockMvcResultMatchers.handler().handlerType(StatisticsController::class.java) }
            .andExpect { MockMvcResultMatchers.handler().methodName("getOttGraph") }
            .andExpect { MockMvcResultMatchers.status().isOk }

        val ottGraphResponse = statisticsService.getOttGraph(1, PeriodOption.ALL)

        val line = ottGraphResponse.ottLineGraph
        val ranking = ottGraphResponse.ottRanking

        line.forEachIndexed { i, ottLineGraphDto ->
            println("Axis Label: ${ottLineGraphDto.axisLabel}, OTT: ${ottLineGraphDto.ottName}, Count: ${ottLineGraphDto.count}")
            resultActions
                .andExpect { MockMvcResultMatchers.jsonPath("$.data.ottLineGraph.[$i].axisLabel").value(ottLineGraphDto.axisLabel) }
                .andExpect { MockMvcResultMatchers.jsonPath("$.data.ottLineGraph.[$i].ottName").value(ottLineGraphDto.ottName) }
                .andExpect { MockMvcResultMatchers.jsonPath("$.data.ottLineGraph.[$i].count").value(ottLineGraphDto.count) }
        }

        ranking.forEachIndexed { i, ottRankDto ->
            println("OTT: ${ottRankDto.ottName}, Count: ${ottRankDto.totalCount}")
            resultActions
                .andExpect { MockMvcResultMatchers.jsonPath("$.data.ottRanking.[$i].ottName").value(ottRankDto.ottName) }
                .andExpect { MockMvcResultMatchers.jsonPath("$.data.ottRanking.[$i].totalCount").value(ottRankDto.totalCount) }
        }
    }
}